compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp

```diff
@@ -1021,10 +1021,107 @@ static DiagnosedSilenceableFailure testMatchCallbackCallback(
   return DiagnosedSilenceableFailure::success();
 }
 
+/// Match callback for a reduction with optional leading and trailing
+/// elementwise operations. Matches *the first* occurrence of such a reduction
+/// within an op associated with the given handle.
+///
+/// Input handles:
+///
+///   - container op, must be associated with one operation.
+///
+/// Output handles:
+///
+///   - leading elementwise op, if any;
+///   - the "fill" op preceding the reduction;
+///   - reduction op;
+///   - trailing elementwise op, if any.
+static DiagnosedSilenceableFailure reductionCallback(
+    transform_dialect::MatchCallbackResult &res, Location loc,
+    const transform::TransformState &state, ValueRange handles) {
+  if (handles.size() != 1 || state.getPayloadOps(handles[0]).size() != 1) {
+    return emitSilenceableFailure(loc)
+           << "expected one handle to one operation";
+  }
+
+  transform_dialect::StructuredOpMatcher pattern, fill, leadingEltwise,
+      trailingEltwise;
+  makeGPUReductionMatcher(pattern, fill, leadingEltwise, trailingEltwise);
+
+  // TODO: need a mechanism for this to go around the entire IR,
+  // potentially with list matches for each group.
+  Operation *root = state.getPayloadOps(handles[0])[0];
+  WalkResult walkResult = root->walk([&](Operation *op) {
```

Comment:
Until this is powerful enough to go over the whole DAG, should we verify something on the root?

Reply:
Verify what exactly?

Reply:
That we're looking at the right piece of IREE IR dispatch.group dispatch.region, variant op or whatever we have at this point in time (I forget offhand) that is unique enough to IREE to guarantee we're in a dispatch.